src/daemon/notifications/popup.vala

```diff
@@ -0,0 +1,353 @@
+				notification: notification
+			);
+		}
+
+		construct {
+			bool has_actions = this.notification.actions.length > 0;
+			bool has_default_action = false;
+
+			// Check for a default action
+			foreach (string action in this.notification.actions) {
+				if (action == "default") {
+					has_default_action = true;
+					break;
+				}
+			}
+
+			// Create the content widgets for the popup
+			var content_box = new Gtk.Box(Gtk.Orientation.VERTICAL, 0) {
+				baseline_position = Gtk.BaselinePosition.CENTER
+			};
+			var contents = new Body(this.notification);
+			this.content_stack.add(content_box);
+			content_box.pack_start(contents, false, true, 0);
+
+			// Hook up the close button
+			contents.Closed.connect(() => {
+				this.Closed(NotificationCloseReason.DISMISSED);
+				this.dismiss();
+			});
+
+			// Add notification actions if any are present
+			if (this.notification.actions.length > 0) {
+				var actions = new ActionBox(this.notification.actions, this.notification.hints.contains("action-icons"));
+				actions.ActionInvoked.connect((action_key) => {
+					this.ActionInvoked(action_key);
+					this.dismiss();
+				});
+				content_box.pack_start(actions, false, true, 0);
+			}
+
+			// Handle mouse enter/leave events to pause/start popup decay
+			this.enter_notify_event.connect(() => {
+				this.stop_decay();
+				return Gdk.EVENT_STOP;
+			});
+
+			this.leave_notify_event.connect(() => {
+				this.begin_decay(this.notification.expire_timeout);
+				return Gdk.EVENT_STOP;
+			});
+
+			// Handle interaction events
+			this.button_release_event.connect(() => {
+				if (has_default_action) {
+					this.ActionInvoked("default");
+				} else if (this.notification.app_info != null && !has_actions) {
+					// Try to launch the application that generated the notification
+					try {
+						notification.app_info.launch(null, null);
+					} catch (Error e) {
+						critical("Unable to launch app: %s", e.message);
+					}
+				}
+
+				// Emit this signal since the notification will be closed to make sure
+				// our latest popup tracking doesn't break.
+				this.Closed(NotificationCloseReason.DISMISSED);
+				this.dismiss();
+				return Gdk.EVENT_STOP;
+			});
+		}
+
+		/**
+		 * Replace the content of this notification with a new notification.
+		 */
+		public void replace(Notification new_notif) {
+			this.stop_decay();
+			var content_box = new Gtk.Box(Gtk.Orientation.VERTICAL, 0) {
+				baseline_position = Gtk.BaselinePosition.CENTER
+			};
+
+			var new_contents = new Body(new_notif);
+			content_box.add(new_contents);
+
+			new_contents.Closed.connect(() => {
+				this.Closed(NotificationCloseReason.DISMISSED);
+				this.dismiss();
+			});
+
+			// Add notification actions if any are present
+			if (new_notif.actions.length > 0) {
+				var actions = new ActionBox(new_notif.actions, new_notif.hints.contains("action-icons"));
+				actions.ActionInvoked.connect((action_key) => {
+					this.ActionInvoked(action_key);
+				});
```

Comment:
Since signature is the same, can we do the following?

Can we just do?

```
actions.ActionInvoked.connect(this.ActionInvoked);
```

Reply:
Same as above.